src/Compilers/CSharp/Portable/Binder/Binder_Await.cs

```diff
@@ -286,7 +302,130 @@ private bool GetAwaitableExpressionInfo(
+                && GetGetResultMethod(getAwaiter, node, expression.Type!, diagnostics, out getResult, out getAwaiterGetResultCall)
+                && (!isRuntimeAsyncEnabled || getRuntimeAwaitAwaiter(awaiterType, out runtimeAsyncAwaitCall));
+
+            bool tryGetRuntimeAwaitHelper(out MethodSymbol? runtimeAwaitHelper)
+            {
+                if (!isRuntimeAsyncEnabled)
+                {
+                    runtimeAwaitHelper = null;
+                    return false;
+                }
+
+                var exprOriginalType = expression.Type!.OriginalDefinition;
+                SpecialMember awaitCall;
+                TypeWithAnnotations? maybeNestedType = null;
+                if (ReferenceEquals(exprOriginalType, GetSpecialType(InternalSpecialType.System_Threading_Tasks_Task, diagnostics, expression.Syntax)))
+                {
+                    awaitCall = SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__AwaitTask;
+                }
+                else if (ReferenceEquals(exprOriginalType, GetSpecialType(InternalSpecialType.System_Threading_Tasks_Task_T, diagnostics, expression.Syntax)))
+                {
+                    awaitCall = SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__AwaitTaskT_T;
+                    maybeNestedType = ((NamedTypeSymbol)expression.Type).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
+                }
+                else if (ReferenceEquals(exprOriginalType, GetSpecialType(InternalSpecialType.System_Threading_Tasks_ValueTask, diagnostics, expression.Syntax)))
+                {
+                    awaitCall = SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__AwaitValueTask;
+                }
+                else if (ReferenceEquals(exprOriginalType, GetSpecialType(InternalSpecialType.System_Threading_Tasks_ValueTask_T, diagnostics, expression.Syntax)))
+                {
+                    awaitCall = SpecialMember.System_Runtime_CompilerServices_AsyncHelpers__AwaitValueTaskT_T;
+                    maybeNestedType = ((NamedTypeSymbol)expression.Type).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0];
+                }
+                else
+                {
+                    runtimeAwaitHelper = null;
+                    return false;
+                }
+
+                runtimeAwaitHelper = (MethodSymbol)GetSpecialTypeMember(awaitCall, diagnostics, expression.Syntax);
+
+                if (runtimeAwaitHelper is null)
+                {
+                    return false;
+                }
+
+                if (maybeNestedType is { } nestedType)
+                {
+                    Debug.Assert(runtimeAwaitHelper.TypeParameters.Length == 1);
+                    runtimeAwaitHelper = runtimeAwaitHelper.Construct([nestedType]);
+                    checkMethodGenericConstraints(runtimeAwaitHelper, diagnostics, expression.Syntax.Location);
+                }
+#if DEBUG
```

Comment:
nit: do we need this directive?

Reply:
I would prefer to keep it, there's no need for the branch in our release code.